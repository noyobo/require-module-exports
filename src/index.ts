import fs from 'node:fs';
import { basename, dirname, extname, join } from 'node:path';
import { buildSync } from 'esbuild';

export function runCjs<T>(filePath: string, code: string) {
  const dir = dirname(filePath);
  const fileName = basename(filePath);
  const ext = extname(fileName);
  const fileBase = `${fileName}.timestamp-${Date.now()}-${Math.random().toString(16).slice(2)}`;
  const tmpFilePath = join(dir, fileBase + ext);

  // TODO: support .mjs

  fs.writeFileSync(tmpFilePath, code, 'utf-8');
  const exports = require(tmpFilePath);
  fs.unlinkSync(tmpFilePath);

  if (exports.__esModule) {
    return exports.default as T;
  }
  return exports as T;
}

/**
 * 读取并运行文件
 */
export const requireModuleExports = (filePath: string) => {
  const result = buildSync({
    entryPoints: [filePath],
    bundle: true,
    minify: true,
    platform: 'node',
    format: 'cjs',
    target: 'es2015',
    sourcemap: false,
    write: false,
  });
  const code = result.outputFiles[0].text;
  return runCjs<Record<string, unknown>>(filePath, code);
};
