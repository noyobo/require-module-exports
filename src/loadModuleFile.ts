import fs from 'node:fs';
import { basename, dirname, extname, join } from 'node:path';
import { buildSync } from 'esbuild';

function requireModule<T>(
  filePath: string,
  code: string,
  exportNames: string | string[] = 'default',
): T {
  const dir = dirname(filePath);
  const fileName = basename(filePath);
  const ext = extname(fileName);
  const fileBase = `${fileName}.timestamp-${Date.now()}-${Math.random().toString(16).slice(2)}`;
  const tmpFilePath = join(dir, fileBase + ext);

  // TODO: support .mjs

  fs.writeFileSync(tmpFilePath, code, 'utf-8');
  const exports = require(tmpFilePath);
  fs.unlinkSync(tmpFilePath);

  if (Array.isArray(exportNames)) {
    return exportNames.reduce((acc, key) => {
      (acc as Record<string, unknown>)[key] = (
        exports as Record<string, unknown>
      )[key];
      return acc;
    }, {} as T);
  }

  if (exportNames === 'default') {
    if (exports.__esModule) {
      return exports.default as T;
    }
  }

  return exports;
}

/**
 * load module from file, and return the exports at node environment
 */
export const loadModuleFile = (
  filePath: string,
  exportNames: string | string[] = 'default',
) => {
  const result = buildSync({
    entryPoints: [filePath],
    bundle: true,
    minify: true,
    platform: 'node',
    format: 'cjs',
    target: 'es2015',
    sourcemap: false,
    write: false,
  });
  const code = result.outputFiles[0].text;
  return requireModule<Record<string, unknown>>(filePath, code, exportNames);
};
