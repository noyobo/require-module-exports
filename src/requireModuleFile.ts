import { buildSync } from 'esbuild';

export function runCjs<T>(
  code: string,
  exportNames: string | string[] = 'default',
): T {
  const module = { exports: {} };
  const fn = new Function('module', 'exports', code);
  fn(module, module.exports);
  const mod = module.exports as T;

  if (Array.isArray(exportNames)) {
    return exportNames.reduce((acc, key) => {
      (acc as Record<string, unknown>)[key] = (mod as Record<string, unknown>)[
        key
      ];
      return acc;
    }, {} as T);
  }

  if (exportNames === 'default') {
    // @ts-ignore
    if (mod?.__esModule) {
      // @ts-ignore
      return mod.default;
    }
  }
  return mod;
}

/**
 * transform module file to commonjs module and run it
 * return the exports object
 */
export const requireModuleFile = (
  filePath: string,
  exportNames: string | string[] = 'default',
) => {
  const result = buildSync({
    entryPoints: [filePath],
    bundle: true,
    minify: true,
    format: 'cjs',
    target: 'es2015',
    sourcemap: false,
    write: false,
  });
  const code = result.outputFiles[0].text;
  return runCjs<Record<string, unknown>>(code, exportNames);
};
