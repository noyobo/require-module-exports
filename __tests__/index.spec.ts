import * as PATH from 'node:path';
import { loadModuleFile, requireModuleFile } from '../src/index';


describe('index', () => {
  describe('require-module-exports', () => {
    it('basic', async () => {
      expect(requireModuleFile(PATH.join(__dirname, './fixtures/a.js'))).toMatchSnapshot();
      expect(loadModuleFile(PATH.join(__dirname, './fixtures/b.ts'))).toMatchSnapshot();
    });

    it('basic export names cjs', async () => {
      expect(requireModuleFile(PATH.join(__dirname, './fixtures/c.js'), ['foo', 'bar'])).toMatchSnapshot();
      expect(requireModuleFile(PATH.join(__dirname, './fixtures/c.js'))).toMatchSnapshot();
    });

    it('basic export names esm', async () => {
      expect(loadModuleFile(PATH.join(__dirname, './fixtures/b.ts'), ['foo', 'bar'])).toMatchSnapshot();
    });
  });
});
