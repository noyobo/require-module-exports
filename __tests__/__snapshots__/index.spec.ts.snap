// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index require-module-exports basic 1`] = `
{
  "a": 1,
}
`;

exports[`index require-module-exports basic 2`] = `
{
  "age": 20,
  "b": "/abc/a.js",
  "name": "b",
  "say": [Function],
}
`;

exports[`index require-module-exports basic export names cjs 1`] = `
{
  "bar": "bar",
  "foo": "foo",
}
`;

exports[`index require-module-exports basic export names cjs 2`] = `
{
  "bar": "bar",
  "baz": "baz",
  "foo": "foo",
}
`;

exports[`index require-module-exports basic export names esm 1`] = `
{
  "bar": "bar",
  "foo": "foo",
}
`;
